File: character_logic/plote.py
"""Plote battle and AI hooks."""

from __future__ import annotations

from typing import Any

from .common import floor_int


CHARACTER_ID = "plote"


def on_action_start_status(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if "화상" not in actor.statuses or not action.is_attack:
        return False
    burn = actor.statuses["화상"]
    damage = floor_int(actor.max_hp * 0.02 * burn.stacks)
    if damage <= 0:
        return False
    battle.fixed_damage(actor, damage, f"화상 {burn.stacks}중첩")
    if target.character_id == CHARACTER_ID:
        recovered = floor_int(burn.stacks * 0.5)
        battle.restore_mp(target, recovered, "영혼 연소")
    return battle.game_over


def apply_condition_effects(battle: Any, choice: Any) -> bool | None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if not action.is_skill(CHARACTER_ID, 2):
        return None
    if battle.record.attack_damage_taken.get(actor.side, 0) > 0:
        return False
    burn = target.statuses.get("화상")
    stacks = burn.stacks if burn else 0
    choice.power = (choice.power or 0) + stacks
    print(f"화상 중첩 수 {stacks}로 위력이 {stacks} 증가했다.")
    return None


def on_hit_after_defense(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_skill(CHARACTER_ID, 0):
        if battle.roll("화상 부여") < 60:
            battle.add_status(target, "화상", 3, 3, actor.name, stack=True)
    elif action.is_skill(CHARACTER_ID, 3):
        battle.add_status(target, "화상", 4, 4, actor.name, stack=True)
        battle.add_cost_effect(target, 1.2, 4, action.name)


def apply_non_attack_effects(battle: Any, choice: Any) -> bool:
    if not choice.action.is_skill(CHARACTER_ID, 1):
        return False
    battle.apply_defense(choice.actor, choice.action.name)
    return True


def on_defense_hit(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    if target.defense_name == "가로막는 불길":
        battle.add_status(actor, "화상", 5, 1, target.name, stack=True)


def ai_score(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    expected_damage: float,
    hit_rate: float,
) -> float:
    value = 0.0
    burn = target.statuses.get("화상")
    burn_stacks = burn.stacks if burn else 0
    incoming = battle.estimate_best_incoming_damage(target, actor)
    counts = battle.recent_kind_counts(target)

    if action.is_skill(CHARACTER_ID, 0) and burn_stacks <= 1:
        value += 160
    if action.is_skill(CHARACTER_ID, 1):
        value += incoming * 0.9 + counts["attack"] * 180
    if action.is_skill(CHARACTER_ID, 2):
        value += burn_stacks * 70
        if burn_stacks > 0 and counts["attack"] <= counts["defense"] + counts["meditation"]:
            value += burn_stacks * 115 + (counts["defense"] + counts["meditation"]) * 360
        if burn_stacks >= 5 and counts["attack"] == 0:
            value += 680
        if incoming > 0 and counts["attack"] >= counts["defense"] + counts["meditation"]:
            value -= min(850, incoming * 16)
        if burn_stacks <= 0:
            value -= 180
    if action.is_skill(CHARACTER_ID, 3):
        value += max(0, 4 - burn_stacks) * 75
    return value

File: web/battle-engine/character-logic/plote.js
"use strict";

const CHARACTER_ID = "plote";

function floorInt(value) {
  return Math.floor(value);
}

function recentKindCounts(battle, fighter) {
  return typeof battle.recentKindCounts === "function"
    ? battle.recentKindCounts(fighter)
    : { attack: 0, defense: 0, meditation: 0 };
}

function addCostEffect(battle, fighter, multiplier, turns, source) {
  if (typeof battle.addCostEffect === "function") {
    battle.addCostEffect(fighter, multiplier, turns, source);
    return;
  }
  const current = fighter.costEffects.find((effect) => effect.source === source);
  if (current) {
    current.multiplier = Number(multiplier);
    current.remaining = Math.max(current.remaining, Number(turns));
  } else {
    fighter.costEffects.push({ multiplier: Number(multiplier), remaining: Number(turns), source });
  }
  battle.logs.push(`${fighter.name}의 액티브 스킬 MP 소모량이 ${turns}턴 동안 x${multiplier}가 된다.`);
}

function onDefenseHit(battle, choice) {
  const actor = choice.actor;
  const target = battle.opponent(actor);
  if (target.defenseName === "가로막는 불길") {
    battle.addStatus(actor, "화상", 5, 1, target.name, true);
  }
}

module.exports = {
  onActionStartStatus(battle, choice) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    const burn = actor.statuses["화상"];
    if (!burn || !choice.action.isAttack) return false;
    const damage = floorInt(actor.maxHp * 0.015 * burn.stacks);
    if (damage <= 0) return false;
    battle.fixedDamage(actor, damage, `화상 ${burn.stacks}중첩`, target);
    if ((battle.activeCharacterId?.(target) || target.characterId) === CHARACTER_ID) {
      battle.restoreMp(target, floorInt(burn.stacks * 0.5), "영혼 연소");
    }
    return battle.gameOver;
  },

  applyConditionEffects(battle, choice) {
    if (!choice.action.isSkill(CHARACTER_ID, 2)) return undefined;
    const actor = choice.actor;
    if (Number(battle.record.attackDamageTaken[actor.side] || 0) > 0) return false;
    const burn = battle.opponent(actor).statuses["화상"];
    const stacks = burn ? Number(burn.stacks) : 0;
    choice.power = Number(choice.power || 0) + stacks;
    battle.logs.push(`화상 중첩 수 ${stacks}로 위력이 ${stacks} 증가했다.`);
    return undefined;
  },

  onHitAfterDefenseAsActor(battle, choice) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    const action = choice.action;
    if (action.isSkill(CHARACTER_ID, 0)) {
      if (battle.roll("화상 부여") < 60) battle.addStatus(target, "화상", 3, 3, actor.name, true);
    } else if (action.isSkill(CHARACTER_ID, 3)) {
      battle.addStatus(target, "화상", 4, 4, actor.name, true);
      addCostEffect(battle, target, 1.2, 4, action.name);
    }
  },

  applyNonAttackEffects(battle, choice) {
    if (!choice.action.isSkill(CHARACTER_ID, 1)) return false;
    battle.applyDefense(choice.actor, choice.action.name);
    return true;
  },

  onDefenseHit,
  onHitAfterDefenseAsTarget: onDefenseHit,

  aiScore(battle, actor, target, action, _expectedDamage, _hitRate) {
    let value = 0;
    const burn = target.statuses["화상"];
    const burnStacks = burn ? Number(burn.stacks) : 0;
    const incoming = battle.estimateBestIncomingDamage(target, actor);
    const counts = recentKindCounts(battle, target);

    if (action.isSkill(CHARACTER_ID, 0) && burnStacks <= 1) value += 160;
    if (action.isSkill(CHARACTER_ID, 1)) {
      value += incoming * 0.9 + Number(counts.attack || 0) * 180;
    }
    if (action.isSkill(CHARACTER_ID, 2)) {
      value += burnStacks * 70;
      if (burnStacks > 0 && Number(counts.attack || 0) <= Number(counts.defense || 0) + Number(counts.meditation || 0)) {
        value += burnStacks * 115 + (Number(counts.defense || 0) + Number(counts.meditation || 0)) * 360;
      }
      if (burnStacks >= 5 && Number(counts.attack || 0) === 0) value += 680;
      if (incoming > 0 && Number(counts.attack || 0) >= Number(counts.defense || 0) + Number(counts.meditation || 0)) {
        value -= Math.min(850, incoming * 16);
      }
      if (burnStacks <= 0) value -= 180;
    }
    if (action.isSkill(CHARACTER_ID, 3)) value += Math.max(0, 4 - burnStacks) * 75;
    return value;
  },
};
